File: pages/LoginPage.js
const { By, until } = require("selenium-webdriver");

class LoginPage {
  constructor(driver) {
    this.driver = driver;
  }

  url = "https://stg-wms.onflow.vn/login";

  usenameInput = By.css("input[placeholder='Nhập email'], input[type='email']");
  passwordInput = By.css(
    "input[placeholder='Nhập mật khẩu'], input[type='password']",
  );
  loginButton = By.css("button[type='submit']");
  continueLoginButton = By.xpath("//button[contains(.,'Tiếp tục đăng nhập')]");
  fcOption = By.xpath("//span[text()='FC HN']");
  confirmFcButton = By.xpath(
    "//button[contains(.,'Bạn đã chọn FC') or contains(.,'Xác nhận')]",
  );
  logoutButton = By.xpath("//button[contains(.,'Đăng xuất')]");
  loginFormLocator = By.xpath(
    "//form[.//input[@placeholder='Nhập email'] or .//input[@placeholder='Nhập mật khẩu']]",
  );
  loginErrorMessageLocator = By.xpath(
    ".//*[contains(@class,'ant-form-item-explain') or contains(@class,'error') or contains(@class,'invalid-feedback') or contains(@role,'alert') or contains(@data-testid,'message')][normalize-space(string())!='']",
  );

  async open() {
    await this.driver.get(this.url);
    await this.waitForLoginForm();
  }

  async waitForLoginForm() {
    const emailField = await this.driver.wait(
      until.elementLocated(this.usenameInput),
      20000,
    );
    await this.driver.wait(until.elementIsVisible(emailField), 20000);
    const passwordField = await this.driver.wait(
      until.elementLocated(this.passwordInput),
      20000,
    );
    await this.driver.wait(until.elementIsVisible(passwordField), 20000);
  }

  async login(email, password) {
    const emailField = await this.driver.findElement(this.usenameInput);
    const passwordField = await this.driver.findElement(this.passwordInput);
    const loginBtn = await this.driver.findElement(this.loginButton);

    await emailField.clear();
    await emailField.sendKeys(email);
    await passwordField.clear();
    await passwordField.sendKeys(password);
    await loginBtn.click();
  }

  async getLoginErrorText() {
    try {
      const form = await this.driver.findElement(this.loginFormLocator);
      const errorElements = await form.findElements(
        this.loginErrorMessageLocator,
      );
      for (const element of errorElements) {
        try {
          if (await element.isDisplayed()) {
            const text = await element.getText();
            if (text && text.trim() && text.trim().length < 300) {
              return text.trim();
            }
          }
        } catch (error) {
          // ignore stale elements or visibility errors
        }
      }
    } catch (error) {
      // fallback to page-wide search if login form is not found
    }

    const fallbackElements = await this.driver.findElements(
      this.loginErrorMessageLocator,
    );
    for (const element of fallbackElements) {
      try {
        if (await element.isDisplayed()) {
          const text = await element.getText();
          if (text && text.trim() && text.trim().length < 300) {
            return text.trim();
          }
        }
      } catch (error) {
        // ignore stale elements or visibility errors
      }
    }
    return "";
  }

  async waitForLoginFailure(timeout = 10000) {
    await this.driver.wait(async () => {
      const currentUrl = await this.driver.getCurrentUrl();
      if (currentUrl.includes("/login")) {
        return true;
      }
      const loginButtons = await this.driver.findElements(this.loginButton);
      return loginButtons.length > 0;
    }, timeout);
  }

  async continueLoginIfNeeded() {
    try {
      const continueBtn = await this.driver.wait(
        until.elementLocated(this.continueLoginButton),
        5000,
      );
      await this.driver.wait(until.elementIsVisible(continueBtn), 5000);
      await continueBtn.click();
    } catch (error) {
      // Continue login modal không xuất hiện, bỏ qua
    }
  }

  async clickElementViaJs(element) {
    await this.driver.executeScript(
      "arguments[0].scrollIntoView({block:'center'}); arguments[0].click();",
      element,
    );
  }

  async selectFc(fcName = "FC HN") {
    const fcOption = By.xpath(`//span[text()='${fcName}']`);
    const warehouse = await this.driver.wait(
      until.elementLocated(fcOption),
      15000,
    );
    await this.driver.wait(until.elementIsVisible(warehouse), 15000);

    try {
      await warehouse.click();
    } catch (error) {
      await this.clickElementViaJs(warehouse);
    }

    const confirmButton = await this.driver.wait(
      until.elementLocated(this.confirmFcButton),
      15000,
    );
    await this.driver.wait(until.elementIsVisible(confirmButton), 15000);

    try {
      await confirmButton.click();
    } catch (error) {
      await this.clickElementViaJs(confirmButton);
    }
  }

  async isLoginFormVisible() {
    try {
      const emailField = await this.driver.findElement(this.usenameInput);
      const passwordField = await this.driver.findElement(this.passwordInput);
      return (
        (await emailField.isDisplayed()) && (await passwordField.isDisplayed())
      );
    } catch (error) {
      return false;
    }
  }

  async waitForLoginSuccess() {
    await this.driver.wait(async () => {
      const currentUrl = await this.driver.getCurrentUrl();
      if (
        currentUrl.includes("/dashboard") ||
        currentUrl.includes("/home") ||
        currentUrl.includes("/wms") ||
        currentUrl.includes("/warehouse")
      ) {
        return true;
      }

      const logoutButtons = await this.driver.findElements(this.logoutButton);
      return logoutButtons.length > 0;
    }, 30000);
  }
}

module.exports = LoginPage;

File: src/test/java/com/example/seleniumtestng/pages/LoginPage.java
package com.example.seleniumtestng.pages;

import com.example.seleniumtestng.config.ConfigReader;
import java.time.Duration;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;

public class LoginPage extends BasePage {
    private final By emailInput = By.cssSelector("input[type='email'], input[placeholder*='email']");
    private final By passwordInput = By.cssSelector("input[type='password']");
    private final By submitButton = By.cssSelector("button[type='submit']");
    private final By continueLoginButton = By.xpath("//button[contains(normalize-space(.),'Tiếp tục đăng nhập')]");
    private final By confirmFcButton = By.xpath("//button[contains(normalize-space(.),'Bạn đã chọn FC') or contains(normalize-space(.),'Xác nhận')]");
    private final By logoutButton = By.xpath("//button[contains(normalize-space(.),'Đăng xuất') or contains(normalize-space(.),'Dang xuat')]");
    private final By errorMessages = By.xpath("//*[contains(@class,'ant-form-item-explain') or contains(@class,'error') or contains(@class,'invalid-feedback') or @role='alert'][normalize-space(string())!='']");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void openWmsLogin() {
        driver.get(ConfigReader.required("WMS_BASE_URL") + "/login");
        waitForLoginForm();
    }

    public void waitForLoginForm() {
        visible(emailInput);
        visible(passwordInput);
    }

    public void login(String email, String password) {
        type(emailInput, email);
        type(passwordInput, password);
        click(submitButton);
    }

    public void continueLoginIfNeeded() {
        try {
            WebElement button = shortWait(5000).until(ExpectedConditions.elementToBeClickable(continueLoginButton));
            try {
                button.click();
            } catch (RuntimeException e) {
                jsClick(button);
            }
            System.out.println("Clicked continue login button");
        } catch (TimeoutException ignored) {
            System.out.println("Continue login modal not shown");
        }
    }

    public void selectFc() {
        String fcName = ConfigReader.getOrDefault("DEFAULT_FC_NAME", "FC HN");
        By fcOption = By.xpath("//span[normalize-space()='" + fcName + "'] | //*[normalize-space()='" + fcName + "']");
        WebElement warehouse = wait.until(ExpectedConditions.visibilityOfElementLocated(fcOption));
        clickWarehouseAndConfirm(warehouse);
    }

    public boolean selectFcIfPresent(long timeoutMillis) {
        String fcName = ConfigReader.getOrDefault("DEFAULT_FC_NAME", "FC HN");
        By fcOption = By.xpath("//span[normalize-space()='" + fcName + "'] | //*[normalize-space()='" + fcName + "']");
        try {
            WebElement warehouse = shortWait(timeoutMillis).until(ExpectedConditions.visibilityOfElementLocated(fcOption));
            clickWarehouseAndConfirm(warehouse);
            return true;
        } catch (RuntimeException ignored) {
            return false;
        }
    }

    public void waitForLoginSuccess() {
        wait.withTimeout(Duration.ofSeconds(30)).until(driver -> {
            String url = driver.getCurrentUrl();
            if (url.contains("/dashboard") || url.contains("/home") || url.contains("/warehouse") || url.contains("/user-setting")) {
                return true;
            }
            return !driver.findElements(logoutButton).isEmpty();
        });
    }

    public void waitForLoginFailure() {
        wait.until(driver -> driver.getCurrentUrl().contains("/login") || isLoginFormVisible());
    }

    public boolean isLoginFormVisible() {
        List<WebElement> emails = all(emailInput);
        List<WebElement> passwords = all(passwordInput);
        return !emails.isEmpty() && !passwords.isEmpty() && emails.get(0).isDisplayed() && passwords.get(0).isDisplayed();
    }

    public String getLoginErrorText() {
        for (WebElement element : all(errorMessages)) {
            try {
                String text = element.getText().trim();
                if (element.isDisplayed() && !text.isEmpty() && text.length() < 300) {
                    return text;
                }
            } catch (RuntimeException ignored) {
            }
        }
        return "";
    }

    private void clickWarehouseAndConfirm(WebElement warehouse) {
        try {
            warehouse.click();
        } catch (RuntimeException e) {
            jsClick(warehouse);
        }

        WebElement confirm = wait.until(ExpectedConditions.visibilityOfElementLocated(confirmFcButton));
        try {
            confirm.click();
        } catch (RuntimeException e) {
            jsClick(confirm);
        }
    }
}
